efi/generate_binary.py

```diff
@@ -66,15 +73,75 @@ def _run_genpeimg(args):
         sys.exit(1)
 
 
+def generate_checksum(data):
+    # This will make sure that the data representing the PE image
+    # is updated with any changes that might have been made by
+    # assigning values to header fields as those are not automatically
+    # updated upon assignment.
+
+    # Get the offset to the CheckSum field in the OptionalHeader
+    # (The offset is the same in PE32 and PE32+)
+    checksum_offset = OPTIONALHDR_CHECKSUM
+
+    checksum = 0
+    # Verify the data is dword-aligned. Add padding if needed
+    #
+    remainder = len(data) % 4
+    data_len = len(data) + ((4 - remainder) * (remainder != 0))
+
+    for i in range(int(data_len / 4)):
+        # Skip the checksum field
+        if i == int(checksum_offset / 4):
+            continue
+        if i + 1 == (int(data_len / 4)) and remainder:
+            dword = struct.unpack("I", data[i * 4 :] + (b"\0" * (4 - remainder)))[0]
+        else:
+            dword = struct.unpack("I", data[i * 4 : i * 4 + 4])[0]
+        # Optimized the calculation (thanks to Emmanuel Bourg for pointing it out!)
+        checksum += dword
+        if checksum >= 2**32:
+            checksum = (checksum & 0xFFFFFFFF) + (checksum >> 32)
+
+    checksum = (checksum & 0xFFFF) + (checksum >> 16)
+    checksum = (checksum) + (checksum >> 16)
+    checksum = checksum & 0xFFFF
+
+    # The length is the one of the original data, not the padded one
+    #
+    return checksum + len(data)
+
+
 def _add_nx_pefile(args):
     # unnecessary if we have genpeimg
     if args.genpeimg:
         return
     try:
         import pefile
     except ImportError:
-        print("Unable to add NX support to binaries without genpeimg or python3-pefile")
-        sys.exit(1)
+        print("Adding NX support manually to the binary")
+        with open(args.outfile, "r+b") as fh:
+            buf = bytearray(fh.read(os.path.getsize(args.outfile)))
+            fh.seek(0)
+            # Set bit 4 (IMAGE_DLLCHARACTERISTICS_NX_COMPAT) of DllCharacteristics within the COFF optional header
+            # DllCharacteristics offset = 0xde size = 2 bytes
+            # https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#optional-header-image-only
+            DllCharacteristics = (
+                buf[OPTIONALHDR_DLLCHARACTERISTICS + 1] << 8
+                | buf[OPTIONALHDR_DLLCHARACTERISTICS]
+            )
```

Comment:
could you use struct.unpack_from perhaps?

Reply:
ah thanks, it's the function I was looking for to replace this ...